Make CacheStorageStore ThreadSafeRefCounted

szewai opened the pull request:
#### 1a37418a03ef643923f804cc27c1c51f18bc3cc7
<pre>
Make CacheStorageStore ThreadSafeRefCounted
<a href="https://bugs.webkit.org/show_bug.cgi?id=282292">https://bugs.webkit.org/show_bug.cgi?id=282292</a>
<a href="https://rdar.apple.com/problem/138874336">rdar://problem/138874336</a>

Reviewed by Chris Dumez.

CacheStorageDiskStore can be accessed from its m_ioQueue and m_callbackQueue, so making it ThreadSafeRefCounted for
safety.

* Source/WebKit/NetworkProcess/storage/CacheStorageStore.h:

Canonical link: <a href="https://commits.webkit.org/285877@main">https://commits.webkit.org/285877@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/314790503d101b8d3bb76ff355a49c1209d1b830

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74095 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53524 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26906 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78453 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25333 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62657 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1309 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/58242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16594 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77162 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48406 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63738 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38652 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45278 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23666 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/66783 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21580 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79987 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1412 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/774 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/66553 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1556 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63756 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/65828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16316 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9751 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/7924 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1376 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/4164 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1405 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1393 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1412 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/NetworkProcess/storage/CacheStorageStore.h:
```diff
@@ -33,7 +33,7 @@ namespace WebKit {
 struct CacheStorageRecordInformation;
 
-class CacheStorageStore : public RefCounted<CacheStorageStore> {
+class CacheStorageStore : public ThreadSafeRefCounted<CacheStorageStore> {
```
What thread should CacheStorageStore be destroyed on? `CacheStorageDiskStore` is a subclass and has String data members. As a result, it cannot be constructed on thread A and be destroyed on thread B. If CacheStorageStore is always expected to be constructed / destroyed on the main runloop, then we'd want to use `DestructionThread::MainRunLoop` as second template parameter to make sure.

szewai replied:
It's constructed on WorkQueue.

cdumez reviewed (approved):
Ok. It seems a bit safer given that this part of the code is multi-threaded. That said, I couldn't find any clearly unsafe use. We normally capture protectedThis in a lambda, pass it to another thread and then WTFMove it back to the original work queue. I didn't see us ref'ing / deref'ing on various threads.

szewai committed:
Make CacheStorageStore ThreadSafeRefCounted
https://bugs.webkit.org/show_bug.cgi?id=282292
rdar://problem/138874336

Reviewed by Chris Dumez.

CacheStorageDiskStore can be accessed from its m_ioQueue and m_callbackQueue, so making it ThreadSafeRefCounted for
safety.

* Source/WebKit/NetworkProcess/storage/CacheStorageStore.h:

Canonical link: https://commits.webkit.org/285877@main
